File: app-android/app/src/main/java/com/disser/kraken/mesh/BleFrameCodec.kt
package com.disser.kraken.mesh

import com.disser.kraken.invite.InvitePayloadCodec
import java.util.Base64
import java.util.UUID
import java.util.zip.CRC32
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString

@Serializable
data class BleFrameChunk(
    @SerialName("frame_version")
    val frameVersion: Int = 1,
    @SerialName("transfer_id")
    val transferId: String,
    @SerialName("sender_peer_id")
    val senderPeerId: String,
    @SerialName("sender_fingerprint")
    val senderFingerprint: String,
    @SerialName("sender_display_name")
    val senderDisplayName: String? = null,
    @SerialName("chunk_index")
    val chunkIndex: Int,
    @SerialName("chunk_count")
    val chunkCount: Int,
    @SerialName("payload_size")
    val payloadSize: Int,
    @SerialName("payload_crc32")
    val payloadCrc32: Long,
    @SerialName("payload_base64")
    val payloadBase64: String,
)

object BleFrameCodec {
    const val MAX_PACKET_BYTES = 32 * 1024
    const val DEFAULT_CHUNK_PAYLOAD_BYTES = 24
    const val MAX_GATT_WRITE_BYTES = 512

    fun encodeChunks(
        packet: KrakenPacket,
        senderPeer: DiscoveredPeer,
        chunkPayloadBytes: Int = DEFAULT_CHUNK_PAYLOAD_BYTES,
    ): List<ByteArray> {
        require(chunkPayloadBytes in 1..1024) { "Invalid BLE chunk payload size." }
        val payload = MeshPacketCodec.encode(packet).encodeToByteArray()
        require(payload.size in 1..MAX_PACKET_BYTES) { "Kraken BLE packet exceeds $MAX_PACKET_BYTES bytes." }
        val transferId = "${packet.packetId}-${UUID.randomUUID()}"
        val crc32 = payload.crc32()
        val chunkCount = (payload.size + chunkPayloadBytes - 1) / chunkPayloadBytes
        return (0 until chunkCount).map { index ->
            val start = index * chunkPayloadBytes
            val end = minOf(start + chunkPayloadBytes, payload.size)
            val chunk = BleFrameChunk(
                transferId = transferId,
                senderPeerId = senderPeer.peerId,
                senderFingerprint = senderPeer.fingerprint,
                senderDisplayName = senderPeer.displayName,
                chunkIndex = index,
                chunkCount = chunkCount,
                payloadSize = payload.size,
                payloadCrc32 = crc32,
                payloadBase64 = Base64.getEncoder().encodeToString(payload.copyOfRange(start, end)),
            )
            InvitePayloadCodec.json.encodeToString(chunk).encodeToByteArray()
        }
    }

    fun decodeChunk(bytes: ByteArray): BleFrameChunk =
        InvitePayloadCodec.json.decodeFromString<BleFrameChunk>(bytes.decodeToString()).validated()

    private fun BleFrameChunk.validated(): BleFrameChunk {
        require(frameVersion == 1) { "Unsupported BLE frame version." }
        require(transferId.isNotBlank()) { "Missing BLE transfer id." }
        require(senderFingerprint.isNotBlank()) { "Missing BLE sender fingerprint." }
        require(chunkCount in 1..512) { "Invalid BLE chunk count." }
        require(chunkIndex in 0 until chunkCount) { "Invalid BLE chunk index." }
        require(payloadSize in 1..MAX_PACKET_BYTES) { "Invalid BLE payload size." }
        Base64.getDecoder().decode(payloadBase64)
        return this
    }

}

private fun ByteArray.crc32(): Long =
    CRC32().also { it.update(this) }.value

class BleFrameReassembler(
    private val clock: () -> Long = { System.currentTimeMillis() },
    private val transferTtlMillis: Long = 60_000,
) {
    private val pending = linkedMapOf<String, PendingTransfer>()

    fun accept(chunk: BleFrameChunk): Result<LanFrameEnvelope?> =
        runCatching {
            pruneExpired()
            val transfer = pending.getOrPut(chunk.transferId) {
                PendingTransfer(
                    firstSeenAtEpochMillis = clock(),
                    senderPeerId = chunk.senderPeerId,
                    senderFingerprint = chunk.senderFingerprint,
                    senderDisplayName = chunk.senderDisplayName,
                    chunkCount = chunk.chunkCount,
                    payloadSize = chunk.payloadSize,
                    payloadCrc32 = chunk.payloadCrc32,
                )
            }
            require(transfer.matches(chunk)) { "BLE transfer metadata mismatch." }
            transfer.chunks[chunk.chunkIndex] = Base64.getDecoder().decode(chunk.payloadBase64)
            if (transfer.chunks.size != transfer.chunkCount) {
                return@runCatching null
            }
            pending.remove(chunk.transferId)
            val payload = ByteArray(transfer.payloadSize)
            var offset = 0
            (0 until transfer.chunkCount).forEach { index ->
                val bytes = requireNotNull(transfer.chunks[index]) { "Missing BLE chunk." }
                require(offset + bytes.size <= payload.size) { "BLE payload exceeds expected size." }
                bytes.copyInto(payload, destinationOffset = offset)
                offset += bytes.size
            }
            require(offset == payload.size) { "BLE payload size mismatch." }
            require(payload.crc32() == transfer.payloadCrc32) { "BLE payload checksum mismatch." }
            val packet = MeshPacketCodec.decode(payload.decodeToString()).getOrElse {
                throw IllegalArgumentException("Malformed BLE packet.", it)
            }
            require(packet.senderFingerprint == transfer.senderFingerprint) {
                "BLE sender fingerprint does not match packet sender."
            }
            LanFrameEnvelope(
                senderPeerId = transfer.senderPeerId,
                senderFingerprint = transfer.senderFingerprint,
                senderDisplayName = transfer.senderDisplayName,
                packet = packet,
            )
        }

    private fun pruneExpired() {
        val now = clock()
        pending.entries.removeIf { (_, transfer) ->
            now - transfer.firstSeenAtEpochMillis > transferTtlMillis
        }
    }

    private data class PendingTransfer(
        val firstSeenAtEpochMillis: Long,
        val senderPeerId: String,
        val senderFingerprint: String,
        val senderDisplayName: String?,
        val chunkCount: Int,
        val payloadSize: Int,
        val payloadCrc32: Long,
        val chunks: MutableMap<Int, ByteArray> = linkedMapOf(),
    ) {
        fun matches(chunk: BleFrameChunk): Boolean =
            senderPeerId == chunk.senderPeerId &&
                senderFingerprint == chunk.senderFingerprint &&
                chunkCount == chunk.chunkCount &&
                payloadSize == chunk.payloadSize &&
                payloadCrc32 == chunk.payloadCrc32
    }
}

File: app-macos/Sources/KrakenDesktopCore/Services/BleFrameCodec.swift
import Foundation

public enum MacBleConstants {
    public static let serviceUuidString = "58a1257c-f4a8-48c8-99d5-917b9863d7c4"
    public static let identityCharacteristicUuidString = "58a1257d-f4a8-48c8-99d5-917b9863d7c4"
    public static let packetCharacteristicUuidString = "58a1257e-f4a8-48c8-99d5-917b9863d7c4"
}

public enum BleFrameCodec {
    public static let maxPacketBytes = 32 * 1024
    public static let defaultChunkPayloadBytes = 24
    public static let maxGattWriteBytes = 512

    public static func encodeChunks(
        packet: KrakenPacket,
        senderPeerId: String,
        senderFingerprint: String,
        senderDisplayName: String? = nil,
        chunkPayloadBytes: Int = defaultChunkPayloadBytes
    ) throws -> [Data] {
        guard (1...1024).contains(chunkPayloadBytes) else {
            throw MacBleTransportError.invalidChunkPayloadSize
        }
        guard packet.senderFingerprint == senderFingerprint else {
            throw MacBleTransportError.senderFingerprintMismatch
        }
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys]
        let payload = try encoder.encode(packet)
        guard payload.count > 0 && payload.count <= maxPacketBytes else {
            throw MacBleTransportError.packetTooLarge
        }
        let transferId = "\(packet.packetId)-\(UUID().uuidString)"
        let crc32 = CRC32.checksum(payload)
        let chunkCount = (payload.count + chunkPayloadBytes - 1) / chunkPayloadBytes

        return try (0..<chunkCount).map { index in
            let start = index * chunkPayloadBytes
            let end = min(start + chunkPayloadBytes, payload.count)
            let chunk = BleFrameChunk(
                transferId: transferId,
                senderPeerId: senderPeerId,
                senderFingerprint: senderFingerprint,
                senderDisplayName: senderDisplayName,
                chunkIndex: index,
                chunkCount: chunkCount,
                payloadSize: payload.count,
                payloadCrc32: crc32,
                payloadBase64: payload[start..<end].base64EncodedString()
            )
            let data = try encoder.encode(chunk)
            guard data.count <= maxGattWriteBytes else {
                throw MacBleTransportError.packetTooLarge
            }
            return data
        }
    }

    public static func decodeChunk(_ data: Data) throws -> BleFrameChunk {
        do {
            let chunk = try JSONDecoder().decode(BleFrameChunk.self, from: data)
            return try validate(chunk)
        } catch let error as MacBleTransportError {
            throw error
        } catch {
            throw MacBleTransportError.malformedChunk
        }
    }

    private static func validate(_ chunk: BleFrameChunk) throws -> BleFrameChunk {
        guard chunk.frameVersion == 1 else {
            throw MacBleTransportError.unsupportedFrameVersion
        }
        guard !chunk.transferId.isEmpty else {
            throw MacBleTransportError.missingTransferId
        }
        guard !chunk.senderFingerprint.isEmpty else {
            throw MacBleTransportError.missingSenderFingerprint
        }
        guard (1...512).contains(chunk.chunkCount) else {
            throw MacBleTransportError.invalidChunkCount
        }
        guard (0..<chunk.chunkCount).contains(chunk.chunkIndex) else {
            throw MacBleTransportError.invalidChunkIndex
        }
        guard (1...maxPacketBytes).contains(chunk.payloadSize) else {
            throw MacBleTransportError.invalidPayloadSize
        }
        guard Data(base64Encoded: chunk.payloadBase64) != nil else {
            throw MacBleTransportError.malformedChunk
        }
        return chunk
    }
}

public final class BleFrameReassembler: @unchecked Sendable {
    private struct PendingTransfer {
        var firstSeenAtEpochMillis: Int64
        var senderPeerId: String
        var senderFingerprint: String
        var senderDisplayName: String?
        var chunkCount: Int
        var payloadSize: Int
        var payloadCrc32: UInt32
        var chunks: [Int: Data] = [:]

        func matches(_ chunk: BleFrameChunk) -> Bool {
            senderPeerId == chunk.senderPeerId &&
                senderFingerprint == chunk.senderFingerprint &&
                chunkCount == chunk.chunkCount &&
                payloadSize == chunk.payloadSize &&
                payloadCrc32 == chunk.payloadCrc32
        }
    }

    private let now: @Sendable () -> Int64
    private let transferTtlMillis: Int64
    private var pending: [String: PendingTransfer] = [:]

    public init(
        now: @escaping @Sendable () -> Int64 = { Int64(Date().timeIntervalSince1970 * 1000) },
        transferTtlMillis: Int64 = 60_000
    ) {
        self.now = now
        self.transferTtlMillis = transferTtlMillis
    }

    public func accept(_ chunk: BleFrameChunk) throws -> LanFrameEnvelope? {
        pruneExpired()
        var transfer = pending[chunk.transferId] ?? PendingTransfer(
            firstSeenAtEpochMillis: now(),
            senderPeerId: chunk.senderPeerId,
            senderFingerprint: chunk.senderFingerprint,
            senderDisplayName: chunk.senderDisplayName,
            chunkCount: chunk.chunkCount,
            payloadSize: chunk.payloadSize,
            payloadCrc32: chunk.payloadCrc32
        )
        guard transfer.matches(chunk) else {
            throw MacBleTransportError.transferMetadataMismatch
        }
        guard let chunkPayload = Data(base64Encoded: chunk.payloadBase64) else {
            throw MacBleTransportError.malformedChunk
        }
        transfer.chunks[chunk.chunkIndex] = chunkPayload
        if transfer.chunks.count != transfer.chunkCount {
            pending[chunk.transferId] = transfer
            return nil
        }
        pending.removeValue(forKey: chunk.transferId)

        var payload = Data()
        for index in 0..<transfer.chunkCount {
            guard let bytes = transfer.chunks[index] else {
                throw MacBleTransportError.missingChunk
            }
            payload.append(bytes)
        }
        guard payload.count == transfer.payloadSize else {
            throw MacBleTransportError.payloadSizeMismatch
        }
        guard CRC32.checksum(payload) == transfer.payloadCrc32 else {
            throw MacBleTransportError.payloadChecksumMismatch
        }
        let packet = try JSONDecoder().decode(KrakenPacket.self, from: payload)
        guard packet.senderFingerprint == transfer.senderFingerprint else {
            throw MacBleTransportError.senderFingerprintMismatch
        }
        return LanFrameEnvelope(
            senderPeerId: transfer.senderPeerId,
            senderFingerprint: transfer.senderFingerprint,
            senderDisplayName: transfer.senderDisplayName,
            packet: packet
        )
    }

    private func pruneExpired() {
        let current = now()
        pending = pending.filter { _, transfer in
            current - transfer.firstSeenAtEpochMillis <= transferTtlMillis
        }
    }
}

private enum CRC32 {
    private static let table: [UInt32] = (0..<256).map { value in
        var crc = UInt32(value)
        for _ in 0..<8 {
            if crc & 1 == 1 {
                crc = 0xEDB88320 ^ (crc >> 1)
            } else {
                crc >>= 1
            }
        }
        return crc
    }

    static func checksum(_ data: Data) -> UInt32 {
        var crc: UInt32 = 0xFFFFFFFF
        for byte in data {
            let index = Int((crc ^ UInt32(byte)) & 0xFF)
            crc = table[index] ^ (crc >> 8)
        }
        return crc ^ 0xFFFFFFFF
    }
}
